src/org/dux/stracetool/Tracer.java

```diff
@@ -1,43 +1,123 @@
 package org.dux.stracetool;
 
-import org.dux.cli.DuxCLI;
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+
+import com.sun.jna.Pointer;
+import com.sun.jna.platform.win32.WinNT;
+import com.sun.jna.platform.win32.Kernel32;
+
+import org.slf4j.LoggerFactory;
 
 import java.io.BufferedReader;
+import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.io.IOException;
+import java.io.File;
+import java.io.FileReader;
+import java.io.PrintStream;
+
+import java.lang.reflect.Field;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
+
+/**
+ * Create with the Tracer.Builder class.
+ */
 
 public class Tracer {
-    // TODO Allow for arbitrary temp file names (need to sanitize in constructor)
-    private static final String TMP_FILE = ".dux_out";
-    private String[] args;
+    public static Logger logger;
+    static {
+        logger = (Logger) LoggerFactory.getLogger(Tracer.class);
+        logger.setLevel(Level.INFO);
+    }
 
-    private static final String[] STRACE_CALL = {
-            "strace",
-            "-o", TMP_FILE               // write to tmp file
+    private static final String[] FILTER_ARGS = {
+            "-e", "trace=open,execve,readlink,fstat,stat,lstat"
     };
 
-    private static final String[] PROCMON_CALL = {
-            "TODO implement"
-    };
+    private List<String> args;
+
+    // Windows only; already incorporated into above args list on Linux
+    private String fileName;
+    private boolean traceSubprocesses;
+    private boolean filterCalls;
+
+    public static class Builder {
+        // Required parameters
+        private final String fileName;
+        private final List<String> traceCommand;
+
+        // Optional parameters
+        private boolean traceSubprocesses = false;
+        private boolean filterCalls = false;
+
+        // The output file will be a CSV regardless of what you name it.
+        public Builder(String fileName, List<String> traceCommand) {
+            // argument checking
+            if (fileName == null || traceCommand == null) {
+                throw new NullPointerException("Arguments cannot be null");
+            }
+            if (fileName.length() == 0) {
+                throw new IllegalArgumentException("File name cannot be empty");
+            }
+            if (traceCommand.size() == 0) {
+                throw new IllegalArgumentException("Trace command cannot be empty");
+            }
+            if (fileName.contains(" ")) {
+                // spaces could mean they are passing arbitrary arguments to
+                // strace or procmon; disallow for now
+                throw new IllegalArgumentException("No spaces allowed in filename");
+            }
+
+            this.fileName = fileName;
+            // TODO How to sanitize this?
```

Comment:
Do we need to sanitize? We're running a program that they could just run themselves without passing it through Dux...

Reply:
Good point. I took this out.